src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3757,78 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
             {
                 Debug.Assert(arguments.Length == 1);
                 learnFromEqualsMethodArguments(node.ReceiverOpt, receiverType, arguments[0], results[0].RValueType);
                 return;
             }
 
-            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, WellKnownMember wellKnownMember)
+            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, TypeSymbol receiverType, WellKnownMember wellKnownMember)
             {
-                var wellKnownMethod = compilation.GetWellKnownTypeMember(wellKnownMember);
-                if (wellKnownMethod is null)
+                var wellKnownMethod = (MethodSymbol)compilation.GetWellKnownTypeMember(wellKnownMember);
+                if (wellKnownMethod is null || receiverType is null)
                 {
                     return false;
                 }
 
                 var wellKnownType = wellKnownMethod.ContainingType;
                 var parameterType = method.Parameters[0].TypeWithAnnotations;
                 var constructedType = wellKnownType.Construct(ImmutableArray.Create(parameterType));
+                var constructedMethod = wellKnownMethod.AsMember(constructedType);
 
-                Symbol constructedMethod = null;
-                foreach (var member in constructedType.GetMembers(WellKnownMemberNames.ObjectEquals))
+                // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
+                if (constructedMethod.Equals(method))
                 {
-                    if (member.OriginalDefinition.Equals(wellKnownMethod))
+                    return true;
+                }
+
+                // check whether 'method', when called on this receiver, is an implementation of 'constructedMethod'.
+                for (var baseType = receiverType; baseType is object && method is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
+                {
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
                     {
-                        constructedMethod = member;
-                        break;
+                        // we know no base type will implement this interface member either
+                        return false;
                     }
-                }
 
-                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
+                    if (implementationMethod.ContainingType.IsInterface)
+                    {
+                        // this doesn't make sense because an interface can only explicitly implement a method from its base interface.
+                        return false;
```

Comment:
>return false; [](start = 24, length = 13)

Do we have a test for this code path?

Reply:
I added some tests related to explicit implementations, including callable explicit implementations compiled from VB, but I wasn't able to think of how to actually hit this return statement.

Reply:
> I added some tests related to explicit implementations, including callable explicit implementations compiled from VB, but I wasn't able to think of how to actually hit this return statement.

Doesn't the test I suggested with a class implementing an interface which implements IEquatable.Equals hit this path?

---
In reply to: [390700717](https://github.com/dotnet/roslyn/pull/41918#discussion_r390700717) [](ancestors = 390700717)

Reply:
This scenario doesn't hit it. Maybe I misunderstood:

```cs
interface I1 : IEquatable<string?>
{
    bool IEquatable<string?>.Equals(string? s) => throw null!;
}

class C : I1
{
    void M1(string? s)
    {
        _ = this.Equals(s)
            ? s.ToString()
            : s.ToString(); // 1
    }
}
```

This is tested in `IEquatableEqualsMethod_DefaultImplementation_02`

Reply:
I also tried an additional scenario in IL:

```il
.class interface public auto ansi abstract I1
    implements class [System.Runtime]System.IEquatable`1<string>
{
    // Methods
    .method public final hidebysig virtual 
    	instance bool MyEquals (
    		string other
    	) cil managed 
    {
    	.override method instance bool class [System.Runtime]System.IEquatable`1<string>::Equals(!0)
    	// Method begins at RVA 0x2050
    	// Code size 2 (0x2)
    	.maxstack 8
    	IL_0000: ldc.i4.0
    	IL_0001: ret
    } // end of method I1::'System.IEquatable<System.String>.Equals'
} // end of class I1
```

```cs
#nullable enable

class C : I1
{
    void M1(string? s)
    {
        _ = this.MyEquals(s)
            ? s.ToString()
            : s.ToString(); // 1
    }
}
```

In this scenario, the default interface method doesn't seem to be callable directly from the deriving class, even though the method is public. changing the interface to a class in the IL seems to make it callable.

Reply:
> This is tested in IEquatableEqualsMethod_DefaultImplementation_02

Does the signature of the Equals method called by M1 match the signature of ```IEquatable<string?>.Equals```?

---
In reply to: [391258729](https://github.com/dotnet/roslyn/pull/41918#discussion_r391258729) [](ancestors = 391258729)

Reply:
> In this scenario, the default interface method doesn't seem to be callable directly from the deriving class, even though the method is public. changing the interface to a class in the IL seems to make it callable.

Making interface implementation callable is not a goal here. The goal is to find an implementation method that is in the interface, ```FindImplementationForInterfaceMember``` doesn't really care whether the method is callable or not.

---
In reply to: [391259461](https://github.com/dotnet/roslyn/pull/41918#discussion_r391259461) [](ancestors = 391259461)

Reply:
Thanks, I think it's clicked for me now. Pushed a new test that hits the code path.